File: platforms/android/stato/src/main/java/com/facebook/stato/core/StatoDynamic.java
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package com.facebook.stato.core;

import javax.annotation.Nullable;
import org.json.JSONArray;
import org.json.JSONObject;

public class StatoDynamic {
  private final Object mObject;

  public StatoDynamic(Object object) {
    mObject = object;
  }

  public @Nullable String asString() {
    if (mObject == null) {
      return null;
    }
    return mObject.toString();
  }

  public int asInt() {
    if (mObject == null) {
      return 0;
    }
    if (mObject instanceof Integer) {
      return (Integer) mObject;
    }
    if (mObject instanceof Long) {
      return ((Long) mObject).intValue();
    }
    if (mObject instanceof Float) {
      return ((Float) mObject).intValue();
    }
    if (mObject instanceof Double) {
      return ((Double) mObject).intValue();
    }
    return 0;
  }

  public long asLong() {
    if (mObject == null) {
      return 0;
    }
    if (mObject instanceof Integer) {
      return (Integer) mObject;
    }
    if (mObject instanceof Long) {
      return (Long) mObject;
    }
    if (mObject instanceof Float) {
      return ((Float) mObject).longValue();
    }
    if (mObject instanceof Double) {
      return ((Double) mObject).longValue();
    }
    return 0;
  }

  public float asFloat() {
    if (mObject == null) {
      return 0;
    }
    if (mObject instanceof Integer) {
      return (Integer) mObject;
    }
    if (mObject instanceof Long) {
      return (Long) mObject;
    }
    if (mObject instanceof Float) {
      return (Float) mObject;
    }
    if (mObject instanceof Double) {
      return ((Double) mObject).floatValue();
    }
    return 0;
  }

  public double asDouble() {
    if (mObject == null) {
      return 0;
    }
    if (mObject instanceof Integer) {
      return (Integer) mObject;
    }
    if (mObject instanceof Long) {
      return (Long) mObject;
    }
    if (mObject instanceof Float) {
      return (Float) mObject;
    }
    if (mObject instanceof Double) {
      return (Double) mObject;
    }
    return 0;
  }

  public boolean asBoolean() {
    if (mObject == null) {
      return false;
    }
    return (Boolean) mObject;
  }

  public StatoObject asObject() {
    if (mObject instanceof JSONObject) {
      return new StatoObject((JSONObject) mObject);
    }
    return (StatoObject) mObject;
  }

  public StatoArray asArray() {
    if (mObject instanceof JSONArray) {
      return new StatoArray((JSONArray) mObject);
    }
    return (StatoArray) mObject;
  }
}

File: platforms/android/android-stato/src/main/kotlin/org/stato/core/StatoDynamic.kt
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package org.stato.core

import org.json.JSONArray
import org.json.JSONObject

class StatoDynamic(private val value: Any?) {

  fun asString(): String {
    return value?.toString() ?: ""
  }

  fun asInt(): Int =
    when (value) {
      is Number -> value.toInt()
      else -> 0
    }


  fun asLong(): Long =
    when (value) {
      is Number -> value.toLong()
      else -> 0
    }

  fun asFloat(): Float =
    when (value) {
      is Number -> value.toFloat()
      else -> 0f
    }

  fun asDouble(): Double =
    when (value) {
      is Number -> value.toDouble()
      else -> 0.0
    }

  fun asBoolean(): Boolean {
    return if (value == null) {
      false
    } else (value as Boolean?)!!
  }

  fun asObject(): StatoObject? = when(value) {
    is JSONObject -> StatoObject(value as JSONObject?)
    else -> value as StatoObject?
    }


  fun asArray(): StatoArray? = when(value) {
    is JSONArray -> StatoArray(value as JSONArray?)
    else -> value as StatoArray?
  }
}
